File: 2023/aoc2023/lib/day17.ex
defmodule Day17 do
  # 1074 too high
  def main do
    grid = read()

    run(grid, :first)
    |> IO.inspect()

    run(grid, :second)
    |> IO.inspect()
  end

  def run(grid, :second) do
    # heat, x, y, conseq, dir
    queue = [{grid[{1, 0}], 1, 0, 1, ">"}, {grid[{0, 1}], 0, 1, 1, "v"}]
    %{minx: _minx, maxx: maxx, miny: _miny, maxy: maxy} = dimensions(grid)
    goal = {maxx, maxy}

    seen =
      Enum.map(queue, fn {h, x, y, conseq, dir} -> {{x, y, conseq, dir}, h} end)
      |> Enum.into(%{})

    run(queue, goal, grid, seen, :second, 0)
  end

  def run(grid, :first) do
    queue = [{grid[{1, 0}], 1, 0, 1, ">"}, {grid[{0, 1}], 0, 1, 1, "v"}]
    %{minx: _minx, maxx: maxx, miny: _miny, maxy: maxy} = dimensions(grid)
    goal = {maxx, maxy}

    seen =
      Enum.map(queue, fn {_h, x, y, conseq, dir} -> {x, y, conseq, dir} end)
      |> Enum.into(MapSet.new())

    run(queue, goal, grid, seen, :first, 0)
  end

  def insert(current, []) do
    [current]
  end

  def insert(current, [prev | rest]) do
    {h, _, _, _, _} = prev
    {heat, _, _, _, _} = current

    cond do
      h >= heat ->
        [current, prev | rest]

      true ->
        [prev | insert(current, rest)]
    end
  end

  def is_valid_dir(">", "<"), do: false
  def is_valid_dir("<", ">"), do: false
  def is_valid_dir("^", "v"), do: false
  def is_valid_dir("v", "^"), do: false
  def is_valid_dir(_, _), do: true

  def get_conseq(current_c, current_dir, new_dir) do
    if new_dir == current_dir do
      current_c + 1
    else
      1
    end
  end

  def can_stop?({_, _, _, conseq, _}, :second), do: conseq >= 4

  def can_stop?(_, :first), do: true

  def run([current | rest], goal, grid, seen, part, depth) do
    {current_heat, current_x, current_y, current_conseq, current_dir} = current

    if {current_x, current_y} == goal and can_stop?(current, part) do
      current_heat
    else
      candidates =
        [
          {current_x + 1, current_y, ">"},
          {current_x - 1, current_y, "<"},
          {current_x, current_y + 1, "v"},
          {current_x, current_y - 1, "^"}
        ]
        |> Enum.filter(fn {x, y, _} -> grid[{x, y}] != nil end)
        |> Enum.map(fn {x, y, dir} ->
          {current_heat + grid[{x, y}], x, y, get_conseq(current_conseq, current_dir, dir), dir}
        end)
        |> Enum.filter(fn {h, x, y, conseq, dir} ->
          if part == :second do
            seen[{x, y, conseq, dir}] == nil or
              seen[{x, y, conseq, dir}] >= h
          else
            not MapSet.member?(seen, {x, y, conseq, dir})
          end
        end)
        |> Enum.filter(fn {_h, _newx, _newy, _new_conseq, new_dir} ->
          if part == :second do
            cond do
              current_conseq < 4 ->
                new_dir == current_dir

              current_conseq == 10 ->
                new_dir != current_dir and is_valid_dir(new_dir, current_dir)

              true ->
                is_valid_dir(new_dir, current_dir)
            end
          else
            if current_conseq == 3 do
              new_dir != current_dir and is_valid_dir(new_dir, current_dir)
            else
              is_valid_dir(new_dir, current_dir)
            end
          end
        end)

      seen =
        if part == :second do
          Enum.reduce(candidates, seen, fn {h, x, y, conseq, dir}, seen ->
            Map.put(seen, {x, y, conseq, dir}, h)
          end)
        else
          Enum.reduce(candidates, seen, fn {_h, x, y, conseq, dir}, seen ->
            MapSet.put(seen, {x, y, conseq, dir})
          end)
        end

      Enum.reduce(candidates, rest, fn c, r -> insert(c, r) end)
      |> run(goal, grid, seen, part, depth + 1)
    end
  end

  def dimensions(grid) do
    minx = Enum.min_by(grid, fn {{x, _y}, _} -> x end) |> elem(0) |> elem(0)
    maxx = Enum.max_by(grid, fn {{x, _y}, _} -> x end) |> elem(0) |> elem(0)
    miny = Enum.min_by(grid, fn {{_x, y}, _} -> y end) |> elem(0) |> elem(1)
    maxy = Enum.max_by(grid, fn {{_x, y}, _} -> y end) |> elem(0) |> elem(1)
    %{minx: minx, maxx: maxx, miny: miny, maxy: maxy}
  end

  def read() do
    File.stream!("data/day17.txt")
    |> Enum.map(fn line ->
      String.trim(line)
      |> String.split("", trim: true)
      |> Enum.with_index()
    end)
    |> Enum.with_index()
    |> Enum.map(fn {row, y} -> Enum.map(row, fn {v, x} -> {{x, y}, String.to_integer(v)} end) end)
    |> Enum.concat()
    |> Enum.into(%{})
  end
end

File: 2023/aoc2023/lib/day17.py
import heapq
from typing import Any, NamedTuple


class Pos(NamedTuple):
    heat: int
    x: int
    y: int
    conseq: int
    dir: str
    prev: Any = None

    def __lt__(self, other):
        return (self.heat, -self.x - self.y) < (other.heat, -other.x - other.y)

    @property
    def key(self):
        return (self.x, self.y, self.conseq, self.dir)


def is_valid(d1, d2):
    return not (
        (d1 == "^" and d2 == "v") or (d1 == "v" and d2 == "^") or (d1 == "<" and d2 == ">") or (d1 == ">" and d2 == "<")
    )


def get_conseq(current_c, current_dir, new_dir):
    if new_dir == current_dir:
        return current_c + 1
    return 1


def p1():
    return solve(filter1, -1, set())


def p2():
    return solve(filter2, 3, {})


def solve(filter_func, min_conseq, seen):
    with open("../data/day17.txt") as f:
        data = [list(map(int, (line.strip()))) for line in f]
    goal = (len(data[0]) - 1, len(data) - 1)

    queue = [
        Pos(heat=data[0][1], x=1, y=0, conseq=1, dir=">", prev=None),
        Pos(heat=data[1][0], x=0, y=1, conseq=1, dir="v", prev=None),
    ]

    if min_conseq == -1:
        seen.add(queue[0].key)
        seen.add(queue[0].key)
    else:
        seen[queue[0].key] = queue[0].heat
        seen[queue[1].key] = queue[1].heat

    heapq.heapify(queue)

    depth = 0
    while queue:
        current = heapq.heappop(queue)
        depth += 1
        if not queue:
            break

        if (current.x, current.y) == goal and current.conseq > min_conseq:
            print(current.heat, depth)
            break

        candidates = [
            (current.x + 1, current.y, ">"),
            (current.x - 1, current.y, "<"),
            (current.x, current.y + 1, "v"),
            (current.x, current.y - 1, "^"),
        ]

        candidates = [c for c in candidates if c[0] < len(data[0]) and c[1] < len(data) and c[0] >= 0 and c[1] >= 0]

        candidates = [
            Pos(
                x=candidate[0],
                y=candidate[1],
                heat=current.heat + data[candidate[1]][candidate[0]],
                conseq=get_conseq(current.conseq, current.dir, candidate[2]),
                dir=candidate[2],
                prev=current,
            )
            for candidate in candidates
        ]
        filtered = filter_func(current, candidates, seen)
        for c in filtered:
            heapq.heappush(queue, c)


def filter2(current, candidates, seen):
    filtered = []
    if current.conseq > 10:
        raise ValueError()
    candidates = [c for c in candidates if c.key not in seen or seen[c.key] >= c.heat]
    if current.conseq < 4:
        for c in candidates:
            if c.dir == current.dir:
                filtered.append(c)
    elif current.conseq == 10:
        for c in candidates:
            if c.dir != current.dir and is_valid(c.dir, current.dir):
                filtered.append(c)
    else:
        for c in candidates:
            if is_valid(c.dir, current.dir):
                filtered.append(c)
    for c in filtered:
        seen[c.key] = c.heat
    return filtered


def filter1(current, candidates, seen):
    candidates = [c for c in candidates if c.key not in seen]
    filtered = []
    if current.conseq == 3:
        for c in candidates:
            if c.dir != current.dir and is_valid(c.dir, current.dir):
                filtered.append(c)
    else:
        for c in candidates:
            if is_valid(c.dir, current.dir):
                filtered.append(c)
    for c in filtered:
        seen.add(c.key)
    return filtered


if __name__ == "__main__":
    p1()
    p2()
